### PR TITLE
Fast batched nonbonded on interaction groups, as fxn of ligand charges and epsilons

### PR DESCRIPTION
**Pull request by maxentile**
Improved Jax utilities for processing batches of snapshots, with the aim of computing a batch of nonbonded energies as an efficient differentiable function of ligand nonbonded parameters.

-------

Main feature:

Utilities for precomputing small summary vectors of a stored trajectory, so that ligand-environment nonbonded interaction group energies as a function of ligand charges (or ligand epsilons) can be computed as dot products with the small summary vectors. This is directly inspired by the linear basis function approach developed by Levi Naden and Michael Shirts in https://pubmed.ncbi.nlm.nih.gov/26580188/ .

(Note: These utilities make reweighting virtually a no-op, cost-wise... Rough timings are ~5 microseconds to compute a reweighted hydration free energy in explicit solvent (using Jax in double-precision CPU mode)...)

-------

Additional refactoring:

* Refactor jax nonbonded functions to use `pairs` ([N, 2] array) instead of a pair of `inds_l`, `inds_r` (2 [N,] arrays)
* Extract function for converting nonbonded exceptions/exclusions to rescale masks from tests.common: `nonbonded.convert_exceptions_to_rescale_masks`
* Extract function to get pairs_from_interaction_groups
* Move unused jax function `rescale_coordinates` to attic
    * (unsure if broken?)
* ~`jax.ops.index_update(x, idx, y)` -> `x.at[idx].set(y)`, `x.at[idx].multiply(0.5)`~
    * This update was already applied in https://github.com/proteneer/timemachine/pull/672 , but in resolving merge conflicts I tweaked these slightly to use named slices (`params.at[:,0].multiply(0.5)` -> `params.at[charge_indices].multiply(0.5)`)
* `np.indices` -> `np.meshgrid` (following https://github.com/proteneer/timemachine/pull/578#discussion_r809613431 )
* Change the return type of reference nonbonded interaction groups (no longer returns pairs, since pairs is now an argument)

-------

Reviewer note: This branch "[vmap-jax-nonbonded-ig](https://github.com/proteneer/timemachine/tree/vmap-jax-nonbonded-ig)" is a scope-reduced version of "[vmap-jax-nonbonded](https://github.com/proteneer/timemachine/tree/vmap-jax-nonbonded)" (omitting complexity of handling intramolecular exceptions / exclusions, etc., for inclusion in a later PR after deboggling week).

### CODE REVIEW

File: tests/test_jax_nonbonded.py [hunk 1]
```diff
@@ -351,10 +367,75 @@ def u_a(x, box, params):
...
+
+    # generate many sets of ligand parameters to test on
+    eps_ligand_0 = epsilons[ligand_idx]
+    q_ligand_0 = charges[ligand_idx]
+
+    for _ in range(50):
+        eps_ligand = np.abs(eps_ligand_0 + (0.2 * onp.random.rand(n_ligand) - 0.1))  # pls don't be negative
+        q_ligand = q_ligand_0 + onp.random.randn(n_ligand)
+
+        expected = u_batch_ref(eps_ligand, q_ligand)
```
Comment (proteneer) [inline tests/test_jax_nonbonded.py:438]: not sure about the use case - but would it be appropriate to check gradients as well?

Comment (maxentile) [reply tests/test_jax_nonbonded.py:438]: use case is reweighting - would be appropriate to check gradients as well

Comment (maxentile) [reply tests/test_jax_nonbonded.py:455]: addressed in https://github.com/proteneer/timemachine/pull/685/commits/0b3c866ecf8e76c00bbb7fe96ee4b2d3aaead79e ?

Comment (proteneer) [reply tests/test_jax_nonbonded.py:455]: thanks!

**Review by proteneer — APPROVED** [review]

**Review by jkausrelay — APPROVED** [review]

**Review by mcwitt — APPROVED** [review]
Looks great! Just had a few minor comments and a documentation request.

File: attic/jax_tricks/rescale.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import numpy as onp
+
+
+def rescale_coordinates(conf, indices, box, scales):
+    """Note: scales unused"""
+
+    mol_sizes = np.expand_dims(onp.bincount(indices), axis=1)
+    mol_centers = jax.ops.segment_sum(conf, indices) / mol_sizes
+
+    new_centers = mol_centers - box[2] * np.floor(np.expand_dims(mol_centers[..., 2], axis=-1) / box[2][2])
```
[the remark this answers is not in the document]
Comment (maxentile) [reply attic/jax_tricks/rescale.py:12]: Looks like an improvement -- will apply!

(Note that I moved this function from `jax_utils.py` to `attic/` without modification, since it looked broken / unused / doesn't reference the scales parameter)

File: tests/test_jax_utils.py [hunk 3]
```diff
@@ -59,22 +59,29 @@ def test_get_all_pairs_indices():
...
 
     for n, m in zip(ns, ms):
-        inds_i, inds_j = get_group_group_indices(n, m)
-        assert (inds_i < n).all()
-        assert (inds_j < m).all()
+        atom_indices = onp.arange(n + m)
+
+        # non-contiguous group indices
+        onp.random.shuffle(atom_indices)
+        group_a_indices, group_b_indices = atom_indices[:n], atom_indices[-m:]
```
[the remark this answers is not in the document]
Comment (maxentile) [reply tests/test_jax_utils.py:78]: Should be the same as long as `(n + m) == len(atom_indices)`, and the suggested expression is more legible

File: timemachine/fe/topology.py [hunk 4]
```diff
@@ -546,8 +547,12 @@ def parameterize_nonbonded(self, ff_q_params, ff_lj_params):
         qlj_params, nb_potential = super().parameterize_nonbonded(ff_q_params, ff_lj_params)
 
         # halve the strength of the charge and the epsilon parameters
-        src_qlj_params = jnp.asarray(qlj_params).at[:, 0].multiply(0.5)
-        src_qlj_params = jnp.asarray(src_qlj_params).at[:, 2].multiply(0.5)
+        charge_indices = jax.ops.index[:, 0]
+        epsilon_indices = jax.ops.index[:, 2]
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/fe/topology.py:551]: Good catch -- will fix!

File: timemachine/potentials/nonbonded.py [hunk 5]
```diff
@@ -15,6 +24,22 @@ def switch_fn(dij, cutoff):
 Array = np.array
 
 
+def combining_rule_sigma(sig_i, sig_j):
+    """Lorentz-Berthelot (sig_i + sig_j) / 2,
+    but assuming sig -> (sig / 2) has been applied
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py:29]: Motivation is mostly for me to stop tripping over `sig_ij = sig_i + sig_j` every few months. (In a few other places -- not just this file -- `sig` means `sig / 2` as an optimization)

File: timemachine/potentials/nonbonded.py [hunk 6]
```diff
@@ -220,15 +262,16 @@ def nonbonded_v3(
...
 
-def nonbonded_v3_on_specific_pairs(conf, params, box, inds_l, inds_r, beta: float, cutoff: Optional[float] = None):
+def nonbonded_v3_on_specific_pairs(conf, params, box, pairs, beta: float, cutoff: Optional[float] = None):
     """See nonbonded_v3 docstring for more details
 
     Notes
     -----
-    * Responsibility of caller to ensure pair indices (inds_l, inds_r) are complete.
-        In case of parameter interpolation, more pairs need to be added.
+    * Responsibility of caller to ensure pair indices are complete.
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py:270]: Ahh, will expand on this. (Had added this comment in response to another PR review comment this this function does not perform a validation of pair indices. The user can pass in an arbitrary collection of pairs, completely omitting some pairs of atoms that are within cutoff, or omitting intramolecular pairs, etc...)

File: attic/jax_tricks/rescale.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcwitt) [reply attic/jax_tricks/rescale.py:12]: ah, sorry, I noticed that this was moved but not that it was moved to `attic/`. No need to modify in that case!

File: timemachine/potentials/nonbonded.py [hunk 7]
```diff
@@ -284,3 +331,119 @@ def validate_coulomb_cutoff(cutoff=1.0, beta=2.0, threshold=1e-2):
...
+    to compute
+    contribution_i = sum_j (q_i * q_j) / d_ij * erfc(beta * d_ij)
+    efficiently as a function of q_i
+
+    precompute
+    prefactor_i = sum_j q_j / d_ij * erfc(beta * d_ij)
+
+    so that
+    contribution_i(q_i) = prefactor_i * q_i
+    """
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py:361]: Agreed -- standardized and improved documentation in https://github.com/proteneer/timemachine/pull/685/commits/b940ab98b5fb0128de2036723dec08bf08ceb6cb

File: tests/test_jax_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply tests/test_jax_utils.py]: Suggestion applied in https://github.com/proteneer/timemachine/pull/685/commits/0ed8175f9f2261d447f1b763c3af497ab649094e !

File: timemachine/fe/topology.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/fe/topology.py]: Suggestion applied in https://github.com/proteneer/timemachine/pull/685/commits/a7f3e2c0d14c50fa3fde81b085b8e096c6b183ee !

File: timemachine/potentials/nonbonded.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py]: Comment expanded in https://github.com/proteneer/timemachine/pull/685/commits/4cdf070c73c3b78578556679ee67d47d1b6ef8b0 !

File: timemachine/potentials/nonbonded.py [hunk 8]
```diff
@@ -284,3 +331,119 @@ def validate_coulomb_cutoff(cutoff=1.0, beta=2.0, threshold=1e-2):
...
+    displacements_ij = jax_utils.delta_r(x_i, x_others, box)
+    d2_ij = np.sum(displacements_ij ** 2, 1)
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py]: Suggestion applied in https://github.com/proteneer/timemachine/pull/685/commits/3fbc03006c2a71662a69d537c18684de1508cca5 !

File: tests/common.py [hunk 9]
```diff
@@ -131,17 +131,7 @@ def prepare_lj_system(
...
-    charge_rescale_mask = np.ones((N, N))
-    for (i, j), exc in zip(exclusion_idxs, scales[:, 0]):
-        charge_rescale_mask[i][j] = 1 - exc
-        charge_rescale_mask[j][i] = 1 - exc
-
-    lj_rescale_mask = np.ones((N, N))
-    for (i, j), exc in zip(exclusion_idxs, scales[:, 1]):
-        lj_rescale_mask[i][j] = 1 - exc
-        lj_rescale_mask[j][i] = 1 - exc
+    charge_rescale_mask, lj_rescale_mask = nonbonded.convert_exceptions_to_rescale_masks(exclusion_idxs, scales, N)
```
Comment (badisa) [inline tests/common.py:134]: Point of clarification, is an exception an exclusion?

File: timemachine/fe/topology.py [hunk 10]
```diff
@@ -546,8 +547,12 @@ def parameterize_nonbonded(self, ff_q_params, ff_lj_params):
         qlj_params, nb_potential = super().parameterize_nonbonded(ff_q_params, ff_lj_params)
 
         # halve the strength of the charge and the epsilon parameters
-        src_qlj_params = jnp.asarray(qlj_params).at[:, 0].multiply(0.5)
-        src_qlj_params = jnp.asarray(src_qlj_params).at[:, 2].multiply(0.5)
+        charge_indices = jax.numpy.index_exp[:, 0]
```
[the remark this answers is not in the document]
Comment (badisa) [reply timemachine/fe/topology.py:550]: Also is there a reason to use the `index_exp` syntax over the `.at` syntax?

File: timemachine/potentials/jax_utils.py [hunk 11]
```diff
@@ -1,34 +1,30 @@
-from typing import Tuple
-
-import jax
 import jax.numpy as np
 import numpy as onp
 from jax import vmap
 
 Array = onp.array
```
Comment (badisa) [inline timemachine/potentials/jax_utils.py:5]: nit: Since you are changing the usage, might be good to switch this to `onp.typing.NDArray`?

File: timemachine/potentials/nonbonded.py [hunk 12]
```diff
@@ -284,3 +333,216 @@ def validate_coulomb_cutoff(cutoff=1.0, beta=2.0, threshold=1e-2):
...
+
+def coulomb_prefactors_on_traj(traj, boxes, charges, ligand_indices, env_indices, beta=2.0, cutoff=np.inf):
+    """Map coulomb_prefactors_on_snapshot over snapshots in a trajectory
+
+    Parameters
+    ----------
+    traj: [T, N, D] array
+    boxes: diagonal [T, D, D] array (or [T] array of Nones)
+    charges: [N] array
+    ligand_indices: [N_lig] array of ints
```
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/potentials/nonbonded.py:401]: They should be disjoint (throw error if not), and ligand_indices should be smaller than env_indices (print warning if not) -- will fix

File: timemachine/fe/topology.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/fe/topology.py:550]: Just to be more explicit -- should have identical behavior

[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/fe/topology.py]: Partially addressed in https://github.com/proteneer/timemachine/pull/685/commits/4becd4a287d2723d6dfb4c3aa8484e93a2b75af3

File: tests/common.py (hunk 9 above)
Comment (maxentile) [reply tests/common.py]: Renamed in https://github.com/proteneer/timemachine/pull/685/commits/c4077aeb359ff1179e6f44d837e610a61ccdfc81

File: timemachine/potentials/jax_utils.py (hunk 11 above)
Comment (maxentile) [reply timemachine/potentials/jax_utils.py]: Updated in https://github.com/proteneer/timemachine/pull/685/commits/78ac174fbfe41e6fccf86917a06aa8c022ae9d7b

File: tests/test_jax_nonbonded.py [hunk 13]
```diff
@@ -351,10 +369,99 @@ def u_a(x, box, params):
...
+        def reweight(eps_ligand, q_ligand):
+            delta_us = (u_batch_fxn(eps_ligand, q_ligand) - u_0) / kBT
+            return one_sided_exp(delta_us)
+
+        return reweight
+
+    reweight_ref = jit(make_reweighter(u_batch_ref))
+    reweight_test = jit(make_reweighter(u_batch_test))
+
+    for _ in range(50):
```
Comment (badisa) [inline tests/test_jax_nonbonded.py:451]: This test takes about 50 seconds on CI, does this need 50 iterations to be correct?

Comment (maxentile) [reply tests/test_jax_nonbonded.py:451]: Ahh, on Friday it was ~6 seconds -- will track down which change made it more expensive and reduce iterations if needed

Comment (maxentile) [reply tests/test_jax_nonbonded.py:451]: Locally ~80-90% of the test time is in computing reference gradients https://github.com/proteneer/timemachine/blob/762fc61401a0a870ea5846c18de46d3bce6ad46f/tests/test_jax_nonbonded.py#L463 , added in response to https://github.com/proteneer/timemachine/pull/685#discussion_r840657495 . 

Will rearrange this slightly and reduce the number of iterations from 50 to 5...

**Review by badisa — APPROVED** [review]

**Comment by maxentile** [comment]
Thanks for the reviews! Enabling auto-merge
